src/persist/src/file.rs

```diff
@@ -181,11 +182,75 @@ impl Persister for FilePersister {
     }
 }
 
+/// Implementation of [Blob] backed by files.
+pub struct MemBlob {
+    base_dir: PathBuf,
+}
+
+impl MemBlob {
+    const LOCKFILE_PATH: &'static str = "LOCK";
+
+    /// Returns a new [MemBlob] which stores files under the given dir.
+    ///
+    /// The contents are `lock_info` are stored in the LOCK file used to prevent
+    /// multiple writers. This should include anything that would help debug an
+    /// unexpected LOCK file, such as version, ip, worker number, etc.
+    pub fn new<P: AsRef<Path>>(base_dir: P, lock_info: &[u8]) -> Result<Self, Error> {
+        fs::create_dir_all(&base_dir)?;
+        {
+            let lockfile_path = base_dir.as_ref().join(Self::LOCKFILE_PATH);
+            let mut lockfile = OpenOptions::new()
+                .write(true)
+                .create_new(true)
+                .open(lockfile_path)?;
+            lockfile.write_all(&lock_info)?;
+            lockfile.sync_all()?;
+        }
+        Ok(MemBlob {
+            base_dir: base_dir.as_ref().to_path_buf(),
+        })
+    }
+
+    fn blob_path(&self, key: &str) -> PathBuf {
+        self.base_dir.join(key)
+    }
+}
+
+impl Blob for MemBlob {
+    fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Error> {
+        let file_path = self.blob_path(key);
+        let mut file = match File::open(file_path) {
+            Ok(file) => file,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(None),
+            Err(err) => return Err(err.into()),
+        };
+        let mut buf = Vec::new();
+        file.read_to_end(&mut buf)?;
+        Ok(Some(buf))
+    }
+
+    fn set(&mut self, key: &str, value: Vec<u8>, allow_overwrite: bool) -> Result<(), Error> {
```

Comment:
it seems like we don't consult the lock file - should we? I'm not sure I'm following how the lock file protects against multiple writers

Reply:
ahh ofc that makes sense thanks!